compiler/src/dotty/tools/dotc/transform/UnrollDefinitions.scala

```diff
@@ -0,0 +1,358 @@
+      val map = new util.EqHashMap[Symbol, ComputedIndicies]
+      _unrolledDefs = map
+      map
+    else
+      local.clear()
+      local
+
+  override def phaseName: String = UnrollDefinitions.name
+
+  override def description: String = UnrollDefinitions.description
+
+  override def changesMembers: Boolean = true
+
+  override def run(using Context): Unit =
+    if ctx.compilationUnit.hasUnrollDefs then
+      super.run // create and run the transformer on the current compilation unit
+
+  def newTransformer(using Context): Transformer =
+    UnrollingTransformer(ctx.compilationUnit.unrolledClasses.nn)
+
+  type ComputedIndicies = List[(Int, List[Int])]
+  type ComputeIndicies = Context ?=> Symbol => ComputedIndicies
+
+  private class UnrollingTransformer(classes: Set[Symbol]) extends Transformer {
+    private val unrolledDefs = initializeUnrolledDefs()
+
+    def computeIndices(annotated: Symbol)(using Context): ComputedIndicies =
+      unrolledDefs.getOrElseUpdate(annotated, {
+        if annotated.name.is(DefaultGetterName) then
+          Nil // happens in curried methods where more than one parameter list has @unroll
+        else
+          val indices = annotated
+            .paramSymss
+            .zipWithIndex
+            .flatMap: (paramClause, paramClauseIndex) =>
+              val annotationIndices = findUnrollAnnotations(paramClause)
+              if (annotationIndices.isEmpty) None
+              else Some((paramClauseIndex, annotationIndices))
+          if indices.nonEmpty then
+            // pre-validation should have occurred in posttyper
+            assert(annotated.is(Final, butNot = Deferred) || annotated.isConstructor || annotated.owner.is(ModuleClass) || annotated.name.is(DefaultGetterName),
+              i"$annotated is not final&concrete, or a constructor")
+          indices
+      })
+    end computeIndices
+
+    override def transform(tree: tpd.Tree)(using Context): tpd.Tree = tree match
+      case tree @ TypeDef(_, impl: Template) if classes(tree.symbol) =>
+        super.transform(cpy.TypeDef(tree)(rhs = unrollTemplate(impl, computeIndices)))
+      case tree =>
+        super.transform(tree)
+  }
+
+  def copyParamSym(sym: Symbol, parent: Symbol)(using Context): (Symbol, Symbol) =
+    val copied = sym.copy(owner = parent, flags = (sym.flags &~ HasDefault), coord = sym.coord)
+    sym -> copied
+
+  def symLocation(sym: Symbol)(using Context) = {
+    val lineDesc =
+      if (sym.span.exists && sym.span != sym.owner.span)
+        s" at line ${sym.srcPos.line + 1}"
+      else ""
+    i"in ${sym.owner}${lineDesc}"
+  }
+
+  def findUnrollAnnotations(params: List[Symbol])(using Context): List[Int] = {
+    params
+      .zipWithIndex
+      .collect {
+        case (v, i) if v.hasAnnotation(defn.UnrollAnnot) =>
+          i
+      }
+  }
+
+  def isTypeClause(p: ParamClause) = p.headOption.exists(_.isInstanceOf[TypeDef])
+
+  def generateSingleForwarder(defdef: DefDef,
```

Comment:
Consider adding a comment to this method to show what it's supposed to generate. Notably what its 8 (!) parameters mean and how they influence the result.